homeassistant/helpers/area_registry.py

```diff
@@ -321,6 +337,30 @@ def _generate_area_id(self, name: str) -> str:
             suggestion = f"{suggestion_base}_{tries}"
         return suggestion
 
+    @callback
+    def _async_setup_cleanup(self) -> None:
+        """Set up the area registry cleanup."""
+        # pylint: disable-next=import-outside-toplevel
+        from . import floor_registry as fr  # Circular dependency
+
+        @callback
+        def _handle_floor_registry_update(event: Event) -> None:
+            """Handle floor registry update.
+
+            Update areas that are associated with a floor that has been removed.
+            """
+            if event.data["action"] != "remove":
+                return
```

Comment:
You can pass this to `async_listen` as an event filter. Although this is already a callback, so not sure if necessary 🤷